### PR TITLE
introduce a `fudge_regions_if_ok` to address false region edges

### LINKED ISSUE #37655: Coercion creates false region inference error
This test case, derived from @sfackler's PR #36340, fails to compile incorrectly:

```rust
trait SliceExt<T> {
    type Item;

    fn get_me<I>(&self, index: I) -> &I::Output
        where I: SliceIndex<Self::Item>;
}

impl<T> SliceExt<T> for [T] {
    type Item = T;

    fn get_me<I>(&self, index: I) -> &I::Output
        where I: SliceIndex<T>
    {
        panic!()
    }
}

pub trait SliceIndex<T> {
    type Output: ?Sized;
}

impl<T> SliceIndex<T> for usize {
    type Output = T;
}

fn foo<'a>(split: &[&'a [u8]]) -> &'a [u8] {
    split.get_me(0)
}

fn main() { }
```

I get the error:

```
lunch-box. rustc --stage0 ~/tmp/issue-36340.rs
error[E0495]: cannot infer an appropriate lifetime for autoref due to conflicting requirements
  --> /home/nmatsakis/tmp/issue-36340.rs:27:11
   |
27 |     split.get_me(0)
   |           ^^^^^^
   |
help: consider using an explicit lifetime parameter as shown: fn foo<'a>(split: &'a [&'a [u8]]) -> &'a [u8]
  --> /home/nmatsakis/tmp/issue-36340.rs:26:1
   |
26 | fn foo<'a>(split: &[&'a [u8]]) -> &'a [u8] {
   | ^

error: aborting due to previous error
```

This is because coercion is going wrong, basically. Note that small changes (such as `&**split.get_me(0)`) work fine.

### PR DESCRIPTION
**Pull request by nikomatsakis**
Fixes #37655.

r? @eddyb 
cc @sfackler

### CODE REVIEW

File: src/librustc/infer/fudge.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    /// use the expected types to guide coercions, but we will still
+    /// type-check the resulting types from those coercions against
+    /// the actual types (`?T`, `Option<?T`) -- and remember that
+    /// after the snapshot is popped, the variable `?T` is no longer
+    /// unified.
+    ///
+    /// Assumptions:
+    /// - no new type variables are created during `f()` (asserted
+    ///   below); this simplifies our logic since we don't have to
+    ///   check for escaping type variables
```
Comment (eddyb) [inline src/librustc/infer/fudge.rs]: This will break down once lazy normalization happens, right? Not that it blocks the PR.

**Comment by alexcrichton** [comment]
er, 

@bors: r=eddyb

**Pull request merged**